### PR TITLE
Make peace between Prevote and Check Quorum

### PR DESCRIPTION
**Pull request by Hoverbear**
We discovered some instances where prevote and check quorum interact
poorly.

This PR introduces an extra test, as well as a refinement found in etcd https://github.com/coreos/etcd/pull/8334/

### CODE REVIEW

**Comment by siddontang** [comment]
I think we can change the test to

```rust
let timeout = network.peers[&3].get_randomized_election_timeout();
    for _ in 0..timeout {
        network.peers.get_mut(&3).unwrap().tick();
    }

    network.send(vec![new_message(2, 2, MessageType::MsgHup, 0)]);
    assert_eq!(
        network.peers[&2].state,
        StateRole::Leader,
        "peer 2 state: {:?}, want {:?}",
        network.peers[&2].state,
        StateRole::PreCandidate,
    );

    assert_eq!(
        network.peers[&3].state,
        StateRole::Follower,
        "peer 3 state: {:?}, want {:?}",
        network.peers[&3].state,
        StateRole::Follower,
    );
```

Let 3 steps and becomes PreCandidate, then it will reject 2's PreVote if we don't reset the leader in `become_precandidate`

The origin test can't cover the "lease is not expired" case.

/cc @lishuai87

File: src/raft.rs [hunk 1]
```diff
@@ -828,6 +828,11 @@ impl<T: Storage> Raft<T> {
         // but doesn't change anything else. In particular it does not increase
         // self.term or change self.vote.
         self.state = StateRole::PreCandidate;
+        // If a network partition happens, and leader is in minority partition,
+        // it will step down, and become follower without notifying others.
+        if self.check_quorum {
```
Comment (BusyJay) [inline src/raft.rs]: Why? I think it make sense to reset it whenever it starts elections.

Comment (Hoverbear) [reply src/raft.rs]: Ok. :+1:

**Comment by lishuai87** [comment]
LGTM

**Pull request merged**